include/zephyr/logging/log.h

```diff
@@ -89,13 +90,351 @@ extern "C" {
  */
 #define LOG_WRN_ONCE(...)					\
 	do {							\
-		static uint8_t __warned;			\
-		if (unlikely(__warned == 0)) {			\
+		static atomic_t __warned;			\
+		if (unlikely(atomic_cas(&__warned, 0, 1))) {	\
 			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);	\
-			__warned = 1;				\
 		}						\
 	} while (0)
 
+#if defined(CONFIG_LOG_RATELIMIT) || defined(__DOXYGEN__)
+/**
+ * @brief Core rate-limited logging macro
+ *
+ * @details Internal macro that provides rate-limited logging functionality.
+ * Uses atomic operations to ensure thread safety in multi-threaded environments.
+ * Only one thread can successfully log a message within the specified rate limit period.
+ *
+ * @param _level Log level (LOG_LEVEL_ERR, LOG_LEVEL_WRN, etc.)
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param _skipped_msg Message to show when messages are skipped.
+ * @param _log_func Logging function to call (Z_LOG or Z_LOG_HEXDUMP)
+ * @param ... Arguments to pass to the logging function.
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, _skipped_msg, _log_func, ...)	\
+	do {									\
+		static atomic_t __last_log_time ;		\
+		static atomic_t __skipped_count ;		\
+		uint32_t __now = k_uptime_get_32();				\
+		uint32_t __last = atomic_get(&__last_log_time);		\
+		uint32_t __diff = __now - __last;				\
+		if (unlikely(__diff >= (_rate_ms))) {			\
```

Comment:
Since this is expected to be called under heavy loops, we can avoid the diff calculation on each loop by precomputing the next time.
```
static  ___next;
uint32_t __now = k_uptime_get_32();
if (unlikely(now > __next)) {
   __next = now + _rate_ms;
    ...
}
```